### PR TITLE
add(scan): Implement `DeleteKeys` `ScanService` request

### LINKED ISSUE #8204: Implement `DeleteKeys` scan service request to delete registered keys and any cached results for those keys
## Motivation

We want to clear cached results and stop scanning for a key with a request to the scan service

## Possible Design

- Send a `ScanTaskCommand::RemoveKeys {}` message to the scanner task
  - Remove the key from the `parsed_keys` collection in the scan task and send a message notifying the caller that it's been removed
- Wait for the scan task to confirm that the key has been removed
- Implement and call a method on Storage to delete results for a key

### PR DESCRIPTION
**Pull request by arya2**
## Motivation

This PR implements the `DeleteKeys` request for `ScanService` to:
- drop keys from the scanner task
- delete keys and their results from the database

Closes #8204.

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [x] Have you added or updated tests?
  - [x] Is the documentation up to date?

##### For significant changes:
  - [x] Is there a summary in the CHANGELOG?
  - [x] Can these changes be split into multiple PRs?

_If a checkbox isn't relevant to the PR, mark it as done._

## Solution

- Updates request/response variant types
- Adds a new argument to `scan::start()` for the scan task channel receiver
- Adds `ScanTask::remove_keys()` method for sending a `RemoveKeys` message
- Adds `ScanTask::process_msgs()` and calls it from the loop in `scan::start()`
- Adds a `delete_sapling_keys()` method to `Storage`
- Calls `remove_keys()` and `delete_sapling_keys()` from the `ScanService` match arm for `Request::DeleteKeys`

### Testing

- Tests that the new `Storage::delete_sapling_results()` method deletes all results for a provided key
- Tests that the ScanService sends a message to the `ScanTask` to remove the keys and that it deletes the results before responding

## Review

Anyone can review. 

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

## Follow Up Work

- Add a test for `ScanTask::process_msgs()`

### CODE REVIEW

File: zebra-scan/src/init.rs [hunk 1]
```diff
@@ -65,10 +71,16 @@ impl ScanTask {
         chain_tip_change: ChainTipChange,
     ) -> Self {
         // TODO: Pass `_cmd_receiver` to `scan::start()` to pass it new keys after it's been spawned
```
Comment (oxarbitrage) [inline zebra-scan/src/init.rs]:
```suggestion
```

Is this TODO outdated ? We are passing `cmd_receiver` to `scan_spawn_init` now.

File: zebra-scan/src/init.rs [hunk 2]
```diff
@@ -23,7 +26,7 @@ pub enum ScanTaskCommand {
         done_tx: oneshot::Sender<()>,
 
         /// Key hashes that are to be removed
-        key_hashes: Vec<()>,
+        keys: Vec<String>,
```
Comment (oxarbitrage) [inline zebra-scan/src/init.rs:26]: Should we use `ViewingKey` introduced in https://github.com/ZcashFoundation/zebra/pull/8198/ and in other places or `String` is good enough ?

File: zebra-scan/src/init.rs [hunk 3]
```diff
@@ -36,25 +39,28 @@ pub enum ScanTaskCommand {
...
-    pub handle: JoinHandle<Result<(), Report>>,
+    pub handle: Arc<JoinHandle<Result<(), Report>>>,
 
     /// Task command channel sender
-    cmd_sender: mpsc::Sender<ScanTaskCommand>,
+    pub cmd_sender: mpsc::Sender<ScanTaskCommand>,
 }
 
 impl ScanTask {
     /// Spawns a new [`ScanTask`] for tests.
```
Comment (oxarbitrage) [inline zebra-scan/src/init.rs:50]: If this is only used in tests i was wondering if we should add a `#[cfg(test)]` or similar to it. I understand this is used in the binary too so it might complicate things to do that now.

File: zebra-scan/src/scan.rs [hunk 4]
```diff
@@ -130,6 +137,56 @@ pub async fn start(
...
+
+                Err(TryRecvError::Empty) => break,
+                Err(TryRecvError::Disconnected) => {
+                    // Return early if the sender has been dropped.
+                    return Err(eyre!("command channel disconnected"));
+                }
+            };
+
+            match cmd {
+                ScanTaskCommand::RemoveKeys { done_tx, keys } => {
```
Comment (oxarbitrage) [inline zebra-scan/src/scan.rs:167]: Should we add a limit on the number of keys users can send for deletion in a single call ? Depending in the use case this limit can be greater so i think this can be configurable but for now a const limit for all the keys that can be in any of this calls will be good enough.

File: zebra-scan/src/service/tests.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
+//! Tests for ScanService.
```
Comment (oxarbitrage) [inline zebra-scan/src/service/tests.rs:1]: This tests are great, thank you creating them. One thing i noticed is that we can run this tests without the `shielded-scan` feature flag. Is this something we might want to change?

File: zebra-scan/src/service/tests.rs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    assert!(
+        scan_service
+            .db
+            .sapling_results(&zec_pages_sapling_efvk)
+            .is_empty(),
+        "all results for this key should have been deleted"
+    );
+
+    Ok(())
+}
```
Comment (oxarbitrage) [inline zebra-scan/src/service/tests.rs:79]: I will like to add more tests or maybe just a `TODO` in this PR to delete some keys but make sure other results and keys remain there. The current test add a key and some results and delete them. We also need to make sure that when we have multiple stuff, only what we want is deleted.

**Review by oxarbitrage — DISMISSED** [review]
I think this looks really good. I left some of comments and questions but i don't think any of those are blockers.

File: zebra-scan/src/init.rs (hunk 2 above)
Comment (arya2) [reply zebra-scan/src/init.rs:26]: I think a string is good enough for now. The db methods assume a constant length for the encoded keys, so we may want to use ids or hashes later. We could also serialize the keys into a fixed size to continue using them as the first part of the key.

File: zebra-scan/src/init.rs (hunk 3 above)
Comment (arya2) [reply zebra-scan/src/init.rs:50]: Fixed in https://github.com/ZcashFoundation/zebra/pull/8217/commits/0e1061d5ebd95e72dd6ed5f69bcd72038489d58e.

Comment (oxarbitrage) [reply zebra-scan/src/init.rs:50]: Thanks, that looks good.

File: zebra-scan/src/service/tests.rs (hunk 5 above)
Comment (arya2) [reply zebra-scan/src/service/tests.rs:1]: I think it's okay for the `zebra-scan` unit tests to always run and for the `shielded-scan` feature to enable integration tests with `zebrad`.

File: zebra-scan/src/service/tests.rs (hunk 6 above)
Comment (arya2) [reply zebra-scan/src/service/tests.rs:79]: Updated the test in https://github.com/ZcashFoundation/zebra/pull/8217/commits/dc4e8dd7548f5e7bb851bb1d6671dc86aadbdb89 and https://github.com/ZcashFoundation/zebra/pull/8217/commits/b406ef78743aaa1b160328d3077041c767735b18 to add results for a second key and check that there all of the expected entries/results are there after the first key has been deleted.

File: zebra-scan/src/scan.rs (hunk 4 above)
Comment (arya2) [reply zebra-scan/src/scan.rs:167]: Fixed in https://github.com/ZcashFoundation/zebra/pull/8217/commits/d35ad90fd2feda7d635bf9c11af99202f3d71617.

**Review by oxarbitrage — APPROVED** [review]
thanks for the changes.

File: zebra-scan/src/service/tests.rs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    let expected_keys = vec![zec_pages_sapling_efvk.clone()];
+    let cmd_handler_fut = tokio::task::spawn_blocking(move || {
+        let Ok(ScanTaskCommand::RemoveKeys { done_tx, keys }) = cmd_receiver.recv() else {
+            panic!("should successfully receive RemoveKeys message");
+        };
+
+        assert_eq!(keys, expected_keys, "keys should match the request keys");
+
+        done_tx.send(()).expect("send should succeed");
+    });
```
Comment (upbqdn) [inline zebra-scan/src/service/tests.rs:59]: ###### Optional Suggestion

This looks great. I wonder if we could use mocked services from `zebra-test`, for example: https://github.com/ZcashFoundation/zebra/blob/main/zebra-test/src/mock_service.rs and https://github.com/ZcashFoundation/zebra/blob/main/zebra-test/src/transcript.rs. I haven't used them yet, but I'll look into it.

File: zebra-scan/src/init.rs [hunk 8]
```diff
@@ -80,18 +89,21 @@ impl ScanTask {
...
-/// TODO: add a test for this function.
-pub fn spawn_init(
-    config: &Config,
-    network: Network,
-    state: scan::State,
-    chain_tip_change: ChainTipChange,
-) -> JoinHandle<Result<(), Report>> {
-    scan::spawn_init(config, network, state, chain_tip_change)
+    /// Sends a message to the scan task to remove the provided viewing keys.
+    pub fn remove_keys(
```
Comment (upbqdn) [inline zebra-scan/src/init.rs:94]: ###### Optional Suggestion

I think it would be good to place this function into a different file since the filename is `init.rs`, but let's do that in a different PR?

File: zebra-scan/src/scan.rs [hunk 9]
```diff
@@ -130,6 +137,56 @@ pub async fn start(
...
+                    // TODO: Replace with Arc::unwrap_or_clone() when it stabilises:
+                    // https://github.com/rust-lang/rust/issues/93610
+                    let mut updated_parsed_keys =
+                        Arc::try_unwrap(parsed_keys).unwrap_or_else(|arc| (*arc).clone());
```
Comment (upbqdn) [inline zebra-scan/src/scan.rs:171]: ###### Optional Suggestion

`Arc::get_mut` might also have been an option here.
